pierone/api.py

```diff
@@ -26,35 +28,97 @@ def __str__(self):
         return 'Unauthorized: token missing or invalid'
 
 
-class DockerImage(collections.namedtuple('DockerImage', 'registry team artifact tag')):
-    @classmethod
-    def parse(cls, image: str):
-        '''
-        >>> DockerImage.parse('x')
-        Traceback (most recent call last):
-        ValueError: Invalid docker image "x" (format must be [REGISTRY/]TEAM/ARTIFACT:TAG)
-        >>> DockerImage.parse('foo/bar')
-        DockerImage(registry=None, team='foo', artifact='bar', tag='')
-        >>> DockerImage.parse('registry/foo/bar:1.9')
-        DockerImage(registry='registry', team='foo', artifact='bar', tag='1.9')
-        '''
-        parts = image.split('/')
-        if len(parts) == 3:
-            registry = parts[0]
-        elif len(parts) < 2:
-            raise ValueError('Invalid docker image "{}" (format must be [REGISTRY/]TEAM/ARTIFACT:TAG)'.format(image))
-        else:
-            registry = None
-        team = parts[-2]
-        artifact, sep, tag = parts[-1].partition(':')
-        return DockerImage(registry=registry, team=team, artifact=artifact, tag=tag)
-
-    def __str__(self):
-        '''
-        >>> str(DockerImage(registry='registry', team='foo', artifact='bar', tag='1.9'))
-        'registry/foo/bar:1.9'
-        '''
-        return '{}/{}/{}:{}'.format(*tuple(self))
+class PierOne:
+
+    def __init__(self, url: str):
+        self.url = url if url.startswith("https://") else "https://" + url
+        self.__access_token = get_token('pierone', ['uid'])
+        self.session = requests.Session()
+        self.session.headers['Authorization'] = 'Bearer {}'.format(self.__access_token)
+
+    def _get(self, path, *args, **kwargs) -> requests.Response:
+        """
+        GETs things from Pier One.
+        """
+        url = self.url + path
+        response = self.session.get(url, *args, **kwargs)
+        response.raise_for_status()
+        return response
+
+    def _post(self, path, json=None, *args, **kwargs) -> requests.Response:
+        """
+        POSTs things to Pier One.
+        """
+        url = self.url + path
+        response = self.session.post(url, json=json, *args, **kwargs)
+        response.raise_for_status()
+        return response
+
+    def get_tag_info(self, image: DockerImage) -> list:
+        """
+        Gets detailed tag information
+        """
+        path = "/teams/{}/artifacts/{}/tags/{}".format(image.team, image.artifact, image.tag)
+
+        try:
+            response = self._get(path)
+        except requests.HTTPError as error:
+            if error.response.status_code == 404:
+                # Raise ArtifactNotFound only if it the image is not found
+                raise ArtifactNotFound(image)
+            raise
+        tag_info = response.json()
+        created_by = tag_info["created_by"]
+        tag_info["created_by"] = KNOWN_USERS.get(created_by, created_by)
```

Comment:
IMHO little comment would not hurt here to give a hint to the reader that the purpose of operation is to replace raw `created_by` usernames with user-friendly ones. `KNOWN_USERS.get(created_by, created_by)` looks suspicious at first.

Reply:
Is it better now?